src/promptflow/promptflow/_sdk/operations/_trace_operations.py

```diff
@@ -34,99 +37,98 @@ def list_line_runs(
-                        attributes[SpanAttributeFieldName.REFERENCED_BATCH_RUN_ID]
-                        + "_"
-                        + attributes[SpanAttributeFieldName.LINE_NUMBER]
-                    )
-                    if referenced_line_run_id not in grouped_orm_spans:
-                        grouped_orm_spans[referenced_line_run_id] = []
-                    grouped_orm_spans[referenced_line_run_id].extend(copy.deepcopy(orm_spans))
+                batch_run_id = obo_attrs[SpanAttributeFieldName.BATCH_RUN_ID]
+                line_number = obo_attrs[SpanAttributeFieldName.LINE_NUMBER]
+                # if traces come from eval flow, use `referenced.batch_run_id` as the batch run id
+                if SpanAttributeFieldName.REFERENCED_BATCH_RUN_ID in obo_attrs:
+                    batch_run_id = obo_attrs[SpanAttributeFieldName.REFERENCED_BATCH_RUN_ID]
+                # batch run line run should consider both batch run id and line number
+                line_run_key = f"{batch_run_id}.{line_number}"
+                aggregated_spans[line_run_key].extend(copy.deepcopy(trace_id_spans))
             else:
-                # others, ignore for now
-                pass
-        for orm_spans in grouped_orm_spans.values():
-            spans = [Span._from_orm_object(orm_span) for orm_span in orm_spans]
-            line_run = LineRun._from_spans(spans)
-            if line_run is not None:
-                line_runs.append(line_run)
-        return line_runs
+                # invalid traces, `LINE_RUN_ID` and `BATCH_RUN_ID` should not appear at the same time
+                warning_message = (
+                    f"Invalid traces found, trace id: {obo_span.trace_id}, "
+                    "`LINE_RUN_ID` and `BATCH_RUN_ID` should not appear at the same time."
+                )
+                _logger.warning(warning_message)
+        # convert dict to list
+        return list(aggregated_spans.values())
 
     def _list_line_runs_with_runs(self, runs: typing.List[str]) -> typing.List[LineRun]:
-        orm_spans = ORMSpan.list_with_runs(runs)
-        # group root spans by lineage:
-        #   1. main + eval
-        #   2. eval
-        grouped_spans = {run: dict() for run in runs}
-        for span in map(Span._from_orm_object, orm_spans):
-            attributes = span._content[SpanFieldName.ATTRIBUTES]
+        orm_spans_group_by_trace_id = ORMLineRun.list_with_runs(runs)
+        # ORM entities to SDK entities
+        spans_group_by_trace_id = list()
+        for orm_spans in orm_spans_group_by_trace_id:
+            spans_group_by_trace_id.append([Span._from_orm_object(orm_span) for orm_span in orm_spans])
+        # aggregation logic is different when runs are specified
+        # so will not call `_aggregate_spans` here
+        grouped_spans = {run: defaultdict(list) for run in runs}
+        for trace_id_spans in spans_group_by_trace_id:
+            obo_span: Span = trace_id_spans[0]
```

Comment:
what's obo stands for?

Reply:
On Behalf Of - the attributes in the list are same, so just pick the first to on behalf of them, there are some comments explaining this within this file, maybe I can short copy some here.